tabletserver: Protect Transaction Pool against hot rows.

michael-berlin opened the pull request:
During BeginExecute() do not start a second transaction if another BeginExecute() for the same row is already in progress. This avoids that updates targeted for a single row can consume all transaction pool slots.

Note that the serialization is bound to the lifetime of the BeginExecute() call and not to the lifetime of the transaction. This made the implementation simpler since no additional plumbing was required. Due to the additional MySQL locking, this should result into two transaction pool slots per row at most. (One pending on COMMIT, one waiting for MySQL in BEGIN+EXECUTE.) If this is not effective enough, we will have to revise this decision.

This change reuses the Consolidator object (but it does not set and reuse the "Result" field of the Consolidator.)

Other change:
- Extended fakesqldb to run a callback before it responds to MySQL. This is required to properly synchronize the concurrent transactions. Without, they may not race and get serialized.

sougou commented on go/vt/tabletserver/txserializer/tx_serializer.go:
```diff
@@ -0,0 +1,220 @@
+	limitPerRowRange int
+
+	mu     sync.Mutex
+	queues map[string]*queue
```
I think this will be simpler if you used a counting semaphores instead: https://github.com/youtube/vitess/blob/master/go/sync2/semaphore.go.

These should be performed while holding lock on mu:
* If map entry is absent, then create a semaphore with limitPerRowRange items (and obtain the first semaphore).
* If item is already present, try getting a semaphore. If it succeeds, you're allowed to wait.
* If you had the lock and are releasing it, return the semaphore. If the semaphore Size is equal to limitPerRowRange, then no one else is waiting. So, delete the entry.

For the waiting part:
* Create a buffered channel of size 1. The creator goroutine has the lock. So, start the transaction. It's as if you added a "token" to the channel and took it back immediately.
* Once done, return the token to the buffered channel.
* Any goroutine that wants to wait (obtained a semaphore) will wait on this channel. When the lock holder releases the "lock" (return the token to the channel), only one of the waiters will get it. That goroutine then becomes the lock holder.

The waiting part is also effectively a counting semaphore, except that you need to explicitly select on the channel to handle canceled contexts.

I think the above scheme should be a lot simpler. Let me know if it makes sense,

michael-berlin replied:
Regarding the waiting part:

The question is whether we want to keep the FIFO order of the requests. I say yes because all requests have a short deadline and by processing them in FIFO order we increase the overall success rate. As far as I understand, the channels approach won't guarantee this as a random Go routine will be picked?

sougou replied:
I don't think any of our pools are fair right now. So, it may not be worth going the extra mile just for this one.

michael-berlin replied:
That doesn't mean we can start doing it ;-)

Overall, I don't think the list.List structure as FIFO is increasing the complexity. In my opinion, instead, it's the tight coupling of the TxSerializer and queue objects which makes the code ugly. But I don't really see a way out if we want to delete a queue object after all requests are done. (And I think we have to delete it because the number of different objects equals the number of accessed rows and is unbounded.) What do you think?

sougou replied:
I started writing a long dissertation. But then I found this: https://github.com/golang/go/issues/11506, which I actually upvoted. I never thought they'd fix it. But they did. So, we get fairness for free :).

michael-berlin commented on go/vt/tabletserver/planbuilder/dml.go:
```diff
@@ -29,6 +29,11 @@ func analyzeUpdate(upd *sqlparser.Update, getTable TableGetter) (plan *ExecPlan,
 
+	// Store the WHERE clause as string for the transaction serializer.
+	buf := sqlparser.NewTrackedBuffer(nil)
+	buf.Myprintf("%v", upd.Where)
```
See the code in tabletserver.go. I'm getting the table name separately. This way I can check for an empty table in the query.

sougou replied:
On second thoughts, I don't think this is safe. If you don't include the table name in WhereClause, you'll start unnecessarily serializing all statements that have the same where clauses, like "where id = 1", which may be against different tables.
You can safely assume that there will always be a table name for DMLs. So, it's better to add it to the string to match.

michael-berlin replied:
See the method `computeTxSerializerKey()` in tabletserver.go.

I get the table name there. I think it's cleaner to store it separately. The txserializer package also uses it separately because the stat vars are only keyed by table name and not by table+WHERE to avoid too many entries in that map.

sougou replied:
nvm. Missed that part.
I was sitting at Starbucks, and typed this up for fun. It compiles, but no idea if it runs: https://github.com/sougou/vitess/commit/71ae9ff2d29e296fefa8bba55a3419589209c8cf.

michael-berlin replied:
Looks very good and much simpler. Thanks! :)

The semaphore basically allowed you to fold the two layers `TxSerializer` and `queue` into one and that makes the locking much easier to reason about. Additionally, it provides a lot of thread safe functionality for free which must not be implemented again.

Let me look into the fairness of this tomorrow and then I'll backport it into this PR. :)

michael-berlin replied:
I've looked into the fairness of waiting Go routines.

If there's any delay and a notion of order among waiting Go routines, they are released in the order they were "queued": https://play.golang.org/p/j4F8g4eCGO
(But if there's no delay, Go routines are woken up in a random order because their queue order was probably random to begin with?)

That's great. It means there's no downside to using only `select` (without a dedicated queue) for our use case.

sougou commented on go/sync2/consolidator.go:
```diff
@@ -29,7 +29,13 @@ func NewConsolidator() *Consolidator {
+	executing sync.RWMutex
+	// pending is the number of requests blocked on this (This includes the
+	// original request which holds the write lock.)
+	pending      int
```
Looks like this is not needed any more. Undo this part?

michael-berlin replied:
Done.

sougou commented on go/cmd/vttablet/vttablet.go:
```diff
@@ -43,12 +43,17 @@ func main() {
 	}
+	if err := tabletenv.VerifyConfig(); err != nil {
+		log.Errorf("invalid config: %v", err)
+		exit.Return(1)
```
We should replace this (as well as the previous return) with log.Fatalf. I think this code predates log.Fatalf support.

I may also be open to performing this check in tabletenv.Init, and it just crashed with log.Fatalf.

michael-berlin replied:
Changing to log.Fatalf sounds fine.

Ideally, I would like to call `VerifyConfig` from `NewServer()` and let it return an error if there's a problem. But I was concerned this would draw in more code changes :)

michael-berlin replied:
The previous code didn't use `log.Fatalf` because our custom wrapper `exit.Return()` was used to exit the main Go routine.

I don't think that's necessary and a regular `os.Exit()`, called via `log.Exit()` is good enough. I've changed it accordingly.

michael-berlin commented:
Sugu, PTAL.

I recommend to use Reviewable to look at the diff since I patched in your commit:

https://reviewable.io/reviews/youtube/vitess/2609#-

For that, select the span between r14 and r25 (click on r14, hold the mouse, drag it to r25 and unhold it.)

Changes are:
- replaced semaphores with simple counters
- readded logging
- added /hotrows HTTP handler (reused Consolidator code for that)
- added unit tests for 100%
- fixed bug that canceled wait did not call unlock()

Once the PR is approved, I'll fix the merge conflict and squash all commits into one.
